rfcs/0148-pipe-operator.md

```diff
@@ -0,0 +1,223 @@
+
+F# has `|>`, called "pipe-forward" operator, with the same semantics.
+Additionally, it also has "pipe-backward" `<|` and `>>`/`<<` for forwards and backwards function composition.
+`<|` is equivalent to function application, however its lower binding order allows removing parentheses:
+`g (f a)` = `g <| f a`. All these operators have the same precedence and are left-associative.
+
+Elm has the same operators as F#.
+
+Haskell has the (backwards) function composition operator `.` in its prelude: `(g . f) a` = `g (f a)`.
+It also has "reverse application" `&`, which is roughly equivalent to `|>`,
+and `$`, which is function application again but right-associative and very weakly binding.
+`.` binds stronger than both.
+
+`|>` is definable as an infix function in several other programming languages,
+and in even more languages as macro or higher-order function (including Nix, that's `lib.pipe`).
+
+# Alternatives
+[alternatives]: #alternatives
+
+For each change this RFC proposes, there is always the trivial alternative of not doing it. See #drawbacks.
+
+## More operators
+
+We could use the occasion and introduce more operators like those mentioned above.
+
+Function composition is mostly interesting for the so-called "point-free" programming style,
+where partially applied compositions of functions are preferred over the introduction of lambda terms.
+However, Nix is not well suited for that programming style for various reasons,
+nor would that point-free style have nearly as many applications in real-world Nixpkgs code.
+
+F#'s reverse-pipe operator has a lot less use due to its left-associativity,
+but a right-associative version of it more similar to Haskell's `$` might be an alternative:
+
+```nix
+defaultPrefsFile = 
+  pkgs.writeText "nixos-default-prefs.js" <|
+  lib.concatStringsSep "\n" <|
+  lib.mapAttrsToList (
+    key: value: ''
+      // ${value.reason}
+      pref("${key}", ${builtins.toJSON value.value});
+    ''
+  )
+  defaultPrefs
+  ;
+```
+
+Adding both pipe directions raises questions about how these two interact when used together.
+F# has the same binding strength for both, but this only works well because both are left-associative.
+Haskell has `&` stronger than `$`, which is very sensible but unlikely to be intuitive to a new user.
+Given that we want to call them `|>` and `<|` instead, then users might equally well to assume both have
+equal strength.
+
+Given these restrictions and the fact that situations where one needs both in Nix are expected to be fairly rare,
+it is recommended to choose either one of `|>` and `<|`, but not have both in the language.
+
+## Change the `pipe` function signature
+
+There are many equivalent ways to declare this function, instead of just using the current design.
+For example, one could flip its arguments to allow a partially-applied point-free style (see above).
+One could also make this a single-argument function so that it only takes the list as argument.
```

Comment:
This suggests that the list is heterogeneous: while most elements must be functions, one of them could be of any type (other elements permitting). Heterogeneous lists would present another unnecessary surprise to casual readers.

Reply:
Another downside of having everything in one list is that one loses the last hint of direction the `pipe` function currently has: by passing the argument first and then the functions, it makes a lot more sense to apply the following functions from top to bottom and not the other way around.